src/fs.register.ts

```diff
@@ -63,40 +62,74 @@ export function eaiAgainBuilder(timeout: (attempt: number) => number): BuildMidd
   return eaiAgain;
 }
 
-const client = new S3Client();
-export const s3Fs = new FsAwsS3V3(client);
-client.middlewareStack.add(
-  eaiAgainBuilder((attempt: number) => 100 + attempt * 1000),
-  { name: 'EAI_AGAIN', step: 'build' },
-);
-client.middlewareStack.add(fqdn, { name: 'FQDN', step: 'finalizeRequest' });
+/**
+ * When a new AWS file system is created copy across the middleware, but only if the middleware does not exist
+ *
+ * @param fsClient Filesystem to setup
+ */
+export function setupS3FileSystem(fsClient: FsAwsS3V3): void {
+  addMiddlewareToS3Client(fsClient.client);
 
-FsAwsS3.MaxListCount = 1000;
-s3Fs.credentials.onFileSystemCreated = (acc: AwsCredentialConfig, fs: FileSystem): void => {
-  logger.debug({ prefix: acc.prefix, roleArn: acc.roleArn }, 'FileSystem:Register');
-
-  if (fs.protocol === 's3') {
-    // TODO this cast can be removed once chunkd is upgraded
-    const fsS3 = fs as FsAwsS3V3;
-    const fsClient = fsS3.s3 as S3LikeV3;
-    fsClient.client.middlewareStack.add(fqdn, { name: 'FQDN', step: 'finalizeRequest' });
+  if (fsClient.credentials == null) return;
+  const oldFind = fsClient.credentials.find.bind(fsClient.credentials);
+  // When file systems are looked up ensure they are registered into `fsa`
+  fsClient.credentials.find = async (path: string): Promise<FsAwsS3 | null> => {
+    const accountConfig = await fsClient.credentials.findCredentials(path);
+    if (accountConfig == null) return null;
+
+    const fileSystem = await oldFind(path);
+    if (fileSystem == null) return null;
+
+    logger.debug({ prefix: path, roleArn: accountConfig.roleArn }, 'FileSystem:Register');
+    fsa.register(accountConfig.prefix, fileSystem);
+    if (fileSystem.s3 && 'client' in fileSystem.s3) {
```

Comment:
Don't we usually use `fileSystem.s3.client != null` for these kinds of checks?

Reply:
When you have a object which you dont know much about you have to do a `"client" in foo` to allow the type checker to narrow the type down.

a minimal example

```typescript
type foo = { bar: 'baz' };

foo.client == null; // Error

'client' in foo && foo.client == null; // Ok
```

This could be reworded to `fileSystem.s3 != null && 'client' in fileSystem.s3` to be more specific though.